### PR TITLE
Enhancements: Added custom exception support and expanded debug_ip_layout m_index to support 16bits

### PR DESCRIPTION
**Pull request by rozumx**
Work Done
---------
+ Added exception class XclBinUtilException()
  - Collects: Exception type, file, line number, function, and exception message
+ Updated return error code (currently default to -1) to now return an error code for a MISSING SECTION (100)
+ Increased the index value in debug_ip_data to now be a 16-bit value.
  - This was accomplished by using a reserved byte in debug_ip_data and extending the index value to use the high byte
  - Updated xclbinutil, xclbincat, and xclbinsplit to construct and deconstruct this debug index value
  - Note: The changes are backwards compatible

### CODE REVIEW

File: src/runtime_src/tools/xclbin/XclBinUtilities.h [hunk 1]
```diff
@@ -40,6 +48,54 @@ std::string format(const std::string& format, Args ... args) {
...
+    const char *function() const throw() {
+      return m_function.c_str();
+    }
+
+    XclBinExceptionType exceptionType() const throw() {
+      return m_eExceptionType;
+    }
+};
+
+#define throw_xclbin_exception(exceptionType, msg) throw XclBinUtilities::XclBinUtilException(exceptionType, msg, __FILE__, __LINE__, __FUNCTION__);
```
Comment (stsoe) [inline src/runtime_src/tools/xclbin/XclBinUtilities.h]: I think code will reader nicer if "throw" is part moved out of the macro and it was called like
"throw xclbin_exception(blah,blah);"

File: src/runtime_src/tools/xclbin/XclBinUtilities.h [hunk 2]
```diff
@@ -40,6 +48,54 @@ std::string format(const std::string& format, Args ... args) {
...
+    : std::runtime_error(_msg)
+    , m_msg(_msg)
+    , m_file(_file)
+    , m_line(_line)
+    , m_function(_function)
+    , m_eExceptionType(_eExceptionType) {
+      // Empty
+    }
+
+    ~XclBinUtilException() throw()
```
[the remark this answers is not in the document]
Comment (rozumx) [reply src/runtime_src/tools/xclbin/XclBinUtilities.h]: Updated code to not use the runtime exception specification.

File: src/runtime_src/tools/xclbin/SectionDebugIPLayout.cxx [hunk 3]
```diff
@@ -148,10 +148,15 @@ SectionDebugIPLayout::marshalToJSON(char* _pDataSection,
...
 
-    XUtil::TRACE(XUtil::format("[%d]: m_type: %d, m_index: %d, m_properties: %d, m_major: %d, m_minor: %d, m_base_address: 0x%lx, m_name: '%s'", 
+    // Reform the index value
+    uint16_t m_virtual_index = (((uint16_t) pHdr->m_debug_ip_data[index].m_index_highbyte) << 8) + (uint16_t) pHdr->m_debug_ip_data[index].m_index_lowbyte;
+
+    XUtil::TRACE(XUtil::format("[%d]: m_type: %d, index: %d (m_index_highbyte: 0x%x, m_index_lowbyte: 0x%x), m_properties: %d, m_major: %d, m_minor: %d, m_base_address: 0x%lx, m_name: '%s'", 
                              index,
                              getDebugIPTypeStr((enum DEBUG_IP_TYPE) pHdr->m_debug_ip_data[index].m_type).c_str(),
-                             (unsigned int) pHdr->m_debug_ip_data[index].m_index,
+                             (unsigned int) m_virtual_index,
```
[the remark this answers is not in the document]
Comment (rozumx) [reply src/runtime_src/tools/xclbin/SectionDebugIPLayout.cxx]: For the SectionDebugIPLayout class, I have updated the code to no longer use c-style casting.  The code has been refactored to use C++ static_cast instead.  

Note: As I work with the other classes to to build xclbinutil, I will be refactoring the code to use the C++ style of casting.

File: src/runtime_src/tools/xclbin/XclBinUtilities.h [hunk 4]
```diff
@@ -40,6 +48,54 @@ std::string format(const std::string& format, Args ... args) {
...
+    , m_eExceptionType(_eExceptionType) {
+      // Empty
+    }
+
+    ~XclBinUtilException() throw() 
+    {
+      // Empty
+    }
+
+    const char *msg() const throw() {
```
[the remark this answers is not in the document]
Comment (rozumx) [reply src/runtime_src/tools/xclbin/XclBinUtilities.h]: Yes it should, rename the method to what().

File: src/runtime_src/tools/xclbin/XclBinUtilities.h (hunk 1 above)
Comment (rozumx) [reply src/runtime_src/tools/xclbin/XclBinUtilities.h]: The code is nicer and I have dropped the macro.  The code has been refactored to throw the XclBinUtilException class (e.g., throw XUtil::XclBinUtilException(XET_MISSING_SECTION, errMsg);)

**Pull request merged**